### ISSUE TITLE
Binary image access to bits - possible garbage data?

### ISSUE THREAD

**Issue by kevingnet** [DanBloomberg/leptonica#298]
Hi, first of all I'd like to congratulate the contributors to this library. It's a great product and I've been sort of studying for a while. I've recently started a personal project in which I use binary images. I've encountered this issue and it's most likely my error, but I'm kind of at a loss and may need help to resolve it. Some background:
- I have a simple binary image created with leptonica and its just "a b c d" characters in black sans font and white background.
- The image is segmented using otus's algorithm and then I get the squares for the letters above, all so far so good.
- Then I want to read the pixels, which in a binary image a byte is 8 pixels by convention. For letters a and b, I get the correct results. However, starting at letter c, I get garbage results and I can't figure out why. 

Following is the hex chars I get while iterating for letter a which is correct as an example and then for letter c to show that what I get when reading strikes me as odd. The image is 128x38 which is deliberate since 128/8 = 16 which in the debugger is the line size so I can see easily in the memory window, I use eclipse.

First some sample output using a bitset<8> and cout <<
Letter a
00000111 11111000 00000000 00000000 
00011111 11111100 00000000 00000000 
00011111 11111110 00000000 00000000 
00011000 00001111 00000000 00000000 
00000000 00000111 10000000 00000000 
00000000 00000011 10000000 00000000 
00000000 00000011 10000000 00000000 
00000011 11111111 10000000 00000000 
00001111 11111111 10000000 00000000 
00011111 11111111 10000000 00000000 
00111110 00000011 10000000 00000000 
00111000 00000011 10000000 00000000 
00111000 00000011 10000000 00000000 
00111000 00000111 10000000 00000000 
00111100 00011111 10000000 00000000 
00011111 11111111 10000000 00000000 
00001111 11111011 10000000 00000000 
00000111 11100011 10000000 00000000 
Letter c
11100000 00000000 00000001 01110001 
11110000 00000000 00000111 01110111 
11110000 00000000 00001111 01111111 
00110000 00000000 00011111 01111110 
00000000 00000000 00011110 01111000 
00000000 00000000 00111100 01111000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111000 01110000 
00000000 00000000 00111100 01111000 
00000000 00000000 00011100 01111000 
00110000 00000000 00011111 01111110 
11110000 00000000 00001111 01111111 
11110000 00000000 00000111 01110111 
11100000 00000000 00000001 01110001 
Notice that the start is incorrect, however chars 3 and 4 (the last two) seem wrong.
it reads backwards so that 4th char is 0x71, 0x01, 0x00, 0xE0, which give the first row above which is not correct, because the space between the c and d chars should be much bigger.

Therefore the issue appears that somehow the data is not read properly because of the closeness between the two said characters. We can see that the space is barely 8 bits, however in the image is at least 24 bits. All 4 letters are separated by a space "a b c d  "
I was trying to read the same with another library but I'd rather not mix them and stick with only one. Also, I know that I can use gray channel, but I'd rather not 'cheat.' I should be able to read binary bits without any issues like I've done in the past. I've already spent about 2 almost 3 days with this issue. I'm probably tired and out of practice with low level images, any help would be really appreciated.

Thanks,
Kevin

**Comment by kevingnet** [comment]
Looks like the issue might be endianness, and the way I'm indexing into the data.

**Comment by kevingnet** [comment]
I think I found the issue, or at least a way to make it easy to process data and it's also faster since it's one operation loop. It is to change to big endian, that way I can access byte by byte in a serial manner.

**Comment by kevingnet** [comment]
That was it! Sorry for the spam, library works I expected, my brain did not.

**Comment by DanBloomberg** [comment]
Hi Kevin,

Glad it worked out for you.  My advice is that you debug your algorithms using images to the largest extent possible.  Go down to 0s and 1s only as a last resort.  Show the images at each step.  Endian problems are immediately obvious if you look at the image.  If you compute bounding boxes, show them superimposed over the images to see what you've got.  That's the special fun of doing image processing: you debug with your eyes  :-)

**Comment by kevingnet** [comment]
It's really a great product and thanks for the advise. I'm thinking about contributing. I've been developing some utilities that I'd like to hear if they might be useful in the library:
1) function to merge boxes of objects that are near up to a x and y threshold, so for example the dot in the 'i' char could be merged
2) function to extract binary pixels copied into a 32 or 64 bit integer, which is what I was having trouble with.

**Comment by DanBloomberg** [comment]
Hi Kevin,

I appreciate your enthusiasm.  The library has about 2700 functions, many dealing with box geometry and pixel depth interconversion (see, e.g., pixconv.c for the latter).

We do a lot of "merging" of components in the image domain using morphology.  It's fast and usually works better than box merging.  There are many examples in the library; for example, prog/pageseg_reg.  Run it as:
     pageseg_reg display

**Comment by kevingnet** [comment]
Hi Dan,
I was kind of suspecting that it already had that functionality, so it's cool. I'll do a more thorough search for that next time. I did find something that may be interesting. The Otsu binarization routine, has a window size of min 16x16 pixels, if the image is smaller than that it may not work. Actually I had a bug that I think the function froze, not 100% sure it was due to that. I'll have to test and see if it needs attention.
Thanks for your feedback.
